apps/ws/src/socket/ws.gateway.ts

```diff
@@ -1,41 +1,113 @@
+        ObservabilityBackgroundTransactionEnum.WS_HANDLE_SOCKET_DISCONNECT,
+        'WS Service',
+        function () {
+          const transaction = nr.getTransaction();
+
+          _this
+            .processDisconnectionRequest(connection)
+            .then(resolve)
+            .catch(reject)
+            .finally(() => {
+              transaction.end();
+            });
+        }
+      );
+    });
+  }
+
   async handleConnection(connection: Socket) {
-    const token = connection.handshake.auth?.token || connection.handshake.query?.token;
+    // eslint-disable-next-line @typescript-eslint/no-this-alias
+    const _this = this;
 
-    if (!token || token === 'null') {
-      return this.disconnect(connection);
-    }
+    return new Promise((resolve, reject) => {
+      nr.startBackgroundTransaction(
+        ObservabilityBackgroundTransactionEnum.WS_HANDLE_SOCKET_CONNECTION,
+        'WS Service',
+        function () {
+          const transaction = nr.getTransaction();
 
+          _this
+            .processConnectionRequest(connection)
+            .then(resolve)
+            .catch(reject)
+            .finally(() => {
+              transaction.end();
+            });
+        }
+      );
+    });
+  }
+
+  private extractToken(connection: Socket): string | undefined {
+    return connection.handshake.auth?.token || connection.handshake.query?.token;
+  }
+
+  private async getSubscriber(token: string): Promise<ISubscriberJwt | undefined> {
     let subscriber: ISubscriberJwt;
 
     try {
       subscriber = await this.jwtService.verify(token as string);
       if (subscriber.aud !== 'widget_user') {
-        return this.disconnect(connection);
+        return;
       }
+
+      return subscriber;
     } catch (e) {
+      return;
+    }
+  }
+
+  private async processDisconnectionRequest(connection: Socket) {
+    const token = this.extractToken(connection);
+
+    if (!token || token === 'null') {
+      return;
+    }
+
+    const subscriber = await this.getSubscriber(token);
+    if (!subscriber) {
+      return;
+    }
+
+    const activeConnections = await this.getActiveConnections(connection, subscriber._id);
+    await this.subscriberOnlineService.handleDisconnection(subscriber, activeConnections);
+  }
+
+  private async processConnectionRequest(connection: Socket) {
+    const token = this.extractToken(connection);
+
+    if (!token || token === 'null') {
+      return this.disconnect(connection);
+    }
+
+    const subscriber = await this.getSubscriber(token);
+    if (!subscriber) {
       return this.disconnect(connection);
     }
 
     await connection.join(subscriber._id);
     await this.subscriberOnlineService.handleConnection(subscriber);
-
-    connection.on('disconnect', async () => {
-      const activeConnections = await this.getActiveConnections(connection, subscriber._id);
-      await this.subscriberOnlineService.handleDisconnection(subscriber, activeConnections);
-    });
```

Comment:
This could potentially lead to a memory hog with all existing running connections